rfcs/2022-04-20-12187-log-namespacing.md

```diff
@@ -0,0 +1,422 @@
+on the event, then it will be placed on the root instead. The placement will be decided for each source independently.
+
+#### Syslog
+
+All data will be placed at the root. Since the structure is known, it's easy to prevent naming collisions here.
+Syslog has a "message" field which might be useful to apply an additional codec to, but this is currently not supported,
+and no support is being added here.
+
+#### Native / Native JSON
+
+When these are used in the "Vector" source, the behavior will not change, and events will be passed through as-is.
+
+### Event Metadata
+
+There are 3 sources of information that will be stored in Event Metadata.
+1. Vector "internal" metadata such as `datadog_api_key`, so it can be read/set by users. This data already exists.
+2. Vector metadata such as `ingest_timestamp`, and `source_type` which are set for every source.
+3. Source metadata which will vary for each source.
+
+There is currently minimal support for event metadata. Several changes will need to be made.
+
+Changes needed immediately:
+
+- Support arbitrarily nested values
+- Functions that access metadata (`get_metadata_field`, `remove_metadata_field`, and `set_metadata_field`) should support full paths as keys, and return
+  the `any` type instead of just `string`. This is technically a breaking change.
+
+With these changes, using metadata can still be a bit annoying since the returned type will always be `any`, even if the
+value is set and read in the same VRL program.
+
+Future changes (out of scope for now):
+
+- Add a special path syntax to reference metadata instead of using functions.
+- Expand schema support to metadata
+- Expand semantic meaning to metadata
+
+
+### Implementation
+
+A proof of concept for the Datadog Agent Logs source is being worked on alongside this RFC: [12218](https://github.com/vectordotdev/vector/pull/12218)
+
+
+### Examples
+
+All examples shown are using the new Vector namespace
+
+#### Datadog Agent source / JSON codec
+
+event
+
+```json
+{
+  "content": {
+    "derivative": -2.266778047142367e+125,
+    "integral": "13028769352377685187",
+    "mineral": "H 9 ",
+    "proportional": 3673342615,
+    "vegetable": -30083
+  },
+  "ddsource": "waters",
+  "ddtags": "env:prod",
+  "hostname": "beta",
+  "service": "cernan",
+  "status": "notice",
+  "timestamp": "2066-08-09T04:24:42.1234Z" // This is parsed from a unix timestamp provided by the DD agent
+}
+```
+
+metadata
+
+```json
+{
+  "log_namespace": "vector",
+  "source_type": "datadog_agent",
+  "ingest_timestamp": "2022-04-14T19:14:21.899623781Z",
+
+  // These are existing fields in event metadata that may be on any event
+  "datadog_api_key": "2o86gyhufa2ugyf4",
+  "splunk_hec_token": "386ygfhawnfud6rjftg"
+}
+```
+
+-----
+
+#### Datadog Agent source / bytes codec
+
+event
+
+```json
+{
+  "content": "{\"proportional\":702036423,\"integral\":15089925750456892008,\"derivative\":-6.4676193438086e263,\"vegetable\":20003,\"mineral\":\"vsd5fwYBv\"}",
```

Comment:
Again, Datadog calls this `message` so I think it'd be confusing not to call it `message` here.